File: src/test/scala/com/exasol/cloudetl/source/AbstractSourceTest.scala
package com.exasol.cloudetl.source

import java.nio.file.Path
import java.nio.file.Paths

import com.exasol.cloudetl.filesystem.FileSystemManager
import com.exasol.cloudetl.parquet.ParquetSourceTest
import com.exasol.cloudetl.storage.FileFormat

import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.fs.FileSystem
import org.scalatest.BeforeAndAfterEach
import org.scalatest.funsuite.AnyFunSuite

class AbstractSourceTest extends AnyFunSuite with BeforeAndAfterEach {

  private[this] var conf: Configuration = _
  private[this] var fileSystem: FileSystem = _

  private[source] val format: String = "dummy"
  private[source] var resourceDir: Path = _

  override final def beforeEach(): Unit = {
    conf = new Configuration()
    fileSystem = FileSystem.get(conf)
    resourceDir = Paths.get(getClass.getResource(s"/data/import/$format").toURI).toAbsolutePath
    ()
  }

  final def getConf(): Configuration = conf

  final def getFileSystem(): FileSystem = fileSystem

  final def getSource(filePath: org.apache.hadoop.fs.Path): Source =
    getSource(filePath, format)

  final def getSource(filePath: org.apache.hadoop.fs.Path, fileFormat: String): Source = {
    val format = FileFormat(fileFormat)
    format match {
      case FileFormat.PARQUET | FileFormat.DELTA => ParquetSourceTest(filePath, conf)
      case _                                     => Source(FileFormat(fileFormat), filePath, conf, fileSystem)
    }
  }

  final def getRecordsCount(filePath: Path): Int = {
    val globbedFilePath = FileSystemManager(fileSystem).getLocalFiles(filePath)
    globbedFilePath.map { file =>
      val src = getSource(file)
      val cnt = src.stream().size
      src.close()
      cnt
    }.sum
  }

}

File: src/test/java/com/exasol/cloudetl/source/AbstractSourceTest.java
package com.exasol.cloudetl.source;

import java.net.URISyntaxException;
import java.nio.file.Path;
import java.nio.file.Paths;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.FileSystem;
import org.junit.jupiter.api.BeforeEach;

import com.exasol.cloudetl.ScalaConverters;
import com.exasol.cloudetl.filesystem.FileSystemManager;
import com.exasol.cloudetl.parquet.ParquetSourceTest;
import com.exasol.cloudetl.storage.FileFormat;

abstract class AbstractSourceTest {
    private Configuration conf;
    private FileSystem fileSystem;
    protected Path resourceDir;

    @BeforeEach
    final void abstractSourceBeforeEach() throws Exception {
        this.conf = new Configuration();
        this.fileSystem = FileSystem.get(this.conf);
        this.resourceDir = Paths.get(getClass().getResource("/data/import/" + getFormat()).toURI()).toAbsolutePath();
    }

    final Configuration getConf() {
        return this.conf;
    }

    final FileSystem getFileSystem() {
        return this.fileSystem;
    }

    final Source getSource(final org.apache.hadoop.fs.Path filePath) {
        return getSource(filePath, getFormat());
    }

    protected String getFormat() {
        return "dummy";
    }

    final Source getSource(final org.apache.hadoop.fs.Path filePath, final String fileFormat) {
        final FileFormat format = FileFormat.apply(fileFormat);
        switch (format) {
        case PARQUET:
        case DELTA:
            return new ParquetSourceTest(filePath, this.conf);
        default:
            return Source.create(format, filePath, this.conf, this.fileSystem);
        }
    }

    final int getRecordsCount(final Path filePath) throws java.io.IOException {
        int sum = 0;
        for (final org.apache.hadoop.fs.Path file : ScalaConverters
                .asJavaList(new FileSystemManager(this.fileSystem).getLocalFiles(filePath))) {
            final Source source = getSource(file);
            try {
                sum += source.stream().size();
            } finally {
                source.close();
            }
        }
        return sum;
    }
}
